packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart

```diff
@@ -0,0 +1,148 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math';
+
+import 'package:process/process.dart';
+
+import '../../base/error_handling_io.dart';
+import '../../base/file_system.dart';
+import '../../base/io.dart';
+import '../../base/logger.dart';
+import '../../flutter_manifest.dart';
+import '../build_system.dart';
+
+/// Applies a series of user-specified asset-transforming packages to an asset file.
+final class AssetTransformer {
+  AssetTransformer({
+    required ProcessManager processManager,
+    required Logger logger,
+    required FileSystem fileSystem,
+    required String dartBinaryPath,
+  })  : _processManager = processManager,
+        _logger = logger,
+        _fileSystem = fileSystem,
+        _dartBinaryPath = dartBinaryPath;
+
+  final ProcessManager _processManager;
+  final Logger _logger;
+  final FileSystem _fileSystem;
+  final String _dartBinaryPath;
+  final Random _random = Random();
+
+  /// The [Source] inputs that targets using this should depend on.
+  ///
+  /// See [Target.inputs].
+  static const List<Source> inputs = <Source>[
+    Source.pattern(
+      '{FLUTTER_ROOT}/packages/flutter_tools/lib/src/build_system/targets/asset_transformer.dart',
+    ),
+  ];
+
+  /// Applies, in sequence, a list of transformers to an [asset] and then copies
+  /// the output to [outputPath].
+  ///
+  /// Returns `true` if successful and `false` otherwise.
+  Future<bool> transformAsset({
+    required File asset,
+    required String outputPath,
+    required String workingDirectory,
+    required List<AssetTransformerEntry> transformerEntries,
+  }) async {
+
+    String getTempFilePath() => '${_random.nextInt(1<<32)}${_fileSystem.path.extension(asset.path)}';
+
+    File tempInputFile = _fileSystem.systemTempDirectory.childFile(getTempFilePath());
```

Comment:
Why do you need to copy this, rather than just ussing `asset`?

Reply:
This defends the original asset file from being modified by the transformer. I could perhaps be convinced that the performance penalty isn't worth it.